### PR TITLE
OBSDOCS-118

### PR DESCRIPTION
**Pull request by smunje1**
[OBSDOCS-118](https://issues.redhat.com/browse/OBSDOCS-118): Configuring JSON log data for Elasticsearch is defined incorrectly
Aligned team: Observability
OCP version for cherry-picking: 
JIRA issues: [OBSDOCS-118](https://issues.redhat.com/browse/OBSDOCS-118)
Preview pages: https://75444--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/logging/log_collection_forwarding/cluster-logging-enabling-json-logging.html#cluster-logging-configuration-of-json-log-data-for-default-elasticsearch_cluster-logging-enabling-json-logging
SME review **completed**: @jcantrill
QE review **completed**: @anpingli 
Peer review **requested**:

### CODE REVIEW

**Review by jcantrill — COMMENTED** [review]
lgtm

**Comment by anpingli** [comment]
line 26 should be  flat_labels.    only Preserve k8s Common Labels can be show in kubernetes.labels. Refer to https://issues.redhat.com/browse/LOG-2388
       "kubernetes": {
            "flat_labels": [
              "logFormat=apache",
              ....
            ]
        }

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 1]
```diff
@@ -79,24 +79,30 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
 [source,yaml]
 ----
-outputDefaults:
+outputDefaults: <1>
  elasticsearch:
-    structuredTypeKey: openshift.labels.myLabel <1>
+    structuredTypeKey: kubernetes.labels.managed <2>
     structuredTypeName: nologformat
+outputs:
+  - name: elasticsearch-secure
```
Comment (anpingli) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:87]: structured Field are defined for default elasicsearch..

Comment (anpingli) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:87]: For  example. 
send json  pod logs  under project-qa-x to index:  app-qa-000000x
send json  pod logs  under project-dev-x to index:  app-dev-000000x
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
metadata:
  name: instance
  namespace: openshift-logging
spec:
  inputs: 
    - name: QELogData
      application:
        namespaces:
        - project-qa1
        - project-qa2
          .....
    - name: DevLogData
      application:
        namespaces:
        - project-dev1
        - project-dev2
          ....
  outputDefaults:
    elasticsearch:
      structuredTypeKey: openshift.labels.mylabel
      structuredTypeName: nologformat
  pipelines:
    - name: pipe_first
      inputRefs:
        - QELogData
      outputRefs:
        - default
      labels:
        mylabel: qa
      parse: json
    - name: pipe_second
      inputRefs:
        - DevLogData
      outputRefs:
        - default
      parse: json
      labels:
        mylabel: dev
```

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 2]
```diff
@@ -79,24 +79,30 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+    - elasticsearch-secure
+    - default
+    parse: json
+    labels:
+      myLabel: myValue <3>
 ----
-<1> Uses the value of the key-value pair that is formed by the OpenShift `myLabel` label.
-<2> The `myLabel` element gives its string value, `myValue`, to the structured log record.
+<1> `outputDefaults` applies only to the default Elasticsearch logstore and not to the other defined Elasticsearch outputs.
+<2> Uses the value of the key-value pair that is formed by the OpenShift `myLabel` label.
```
Comment (jcantrill) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:104]: the ... 'managed'... label

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 3]
```diff
@@ -79,24 +79,30 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+<3> The `myLabel` element gives its string value, `myValue`, to the structured log record.
```
Comment (jcantrill) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: The value of "mylabel" is used to construct the index name where the logs will be written.  This example is no longer valid since on line 84 it was changed from kubernetes to openshift.  The structureFieldKey references a field in the log record, not the ClusterLogForwarder.  Further reviewing this point in the notes, this comment does not make any sense as it is unrelated completely to what is needed to be conveyed for this feature.

**Review by jcantrill — CHANGES_REQUESTED** [review]

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 3 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: 1) > This example is no longer valid since on line 84 it was changed from kubernetes to openshift.

Just to confirm, do we mean to say from openshift to kubernetes?

2) > The value of "mylabel" is used to construct the index name where the logs will be written. This example is no longer valid since on line 84 it was changed from kubernetes to openshift. The structureFieldKey references a field in the log record, not the ClusterLogForwarder. Further reviewing this point in the notes, this comment does not make any sense as it is unrelated completely to what is needed to be conveyed for this feature.

Could you please confirm if the **entire** example needs to be changed here? If yes, could you help with the example that we can update in our docs?

Thank you so much!

Cc anpingli

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 1 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:87]: Made updates as discussed in slack thread: https://redhat-internal.slack.com/archives/GGUR75P60/p1718034341293289

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 2 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:104]: Made updates as discussed in slack thread: https://redhat-internal.slack.com/archives/GGUR75P60/p1718034341293289

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 3 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: Made updates as discussed in slack thread: https://redhat-internal.slack.com/archives/GGUR75P60/p1718034341293289

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 4]
```diff
@@ -79,32 +79,36 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+    parse: json
 ----
-<1> Uses the value of the key-value pair that is formed by the OpenShift `myLabel` label.
-<2> The `myLabel` element gives its string value, `myValue`, to the structured log record.
+<1> `outputDefaults` applies only to the default Elasticsearch logstore and not to the other defined Elasticsearch outputs.
+<2> Use the value found at this path in the record to format the index name for the default output.
+<3> Use the value found at this path in the record to format the index name for the elasticsearch output named `elasticsearch-secure`.
 
-In that case, the following structured log record goes to the `app-myValue-write` index:
+In that case, the following structured log record goes to the `app-unmanaged-write` index:
```
Comment (jcantrill) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: This would be "myvalue" since ".openshift.labels.unmanaged" = "myValue"

**Review by jcantrill — COMMENTED** [review]
/lgtm

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 5]
```diff
@@ -79,32 +79,36 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
 
 [source,yaml]
 ----
```
Comment (anpingli) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:81]: outputs:
  - name: default
    elasticsearch:
      structuredTypeKey: openshift.labels.test
      structuredTypeName: nologformat
pipelines:
  - name: application-logs
    inputRefs:
    - application
    - audit
    outputRefs:
    - elasticsearch-secure
    - default
    parse: json
    labels:
      test: test

Comment (anpingli) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:81]: openshift.labels.test is the label defined in pipeline.  not in the source code.

Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:81]: Made updates as discussed in slack thread: https://redhat-internal.slack.com/archives/GGUR75P60/p1718034341293289

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 6]
```diff
@@ -79,32 +79,38 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+      unmanaged: myvalue
 ----
-<1> Uses the value of the key-value pair that is formed by the OpenShift `myLabel` label.
-<2> The `myLabel` element gives its string value, `myValue`, to the structured log record.
+<1> `outputDefaults` applies only to the default Elasticsearch logstore and not to the other defined Elasticsearch outputs.
+<2> Use the value found at this path in the record to format the index name for the default output.
+<3> Use the value defined in pipeline to format the index name for the elasticsearch output named `elasticsearch-secure`.
 
-In that case, the following structured log record goes to the `app-myValue-write` index:
+In that case, the following structured log record goes to the `app-myvalue-write` index:
```
Comment (xenolinux) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:107]: "following structured log record goes to the ..." reads strange. Can you rephrase this? 

Something like this maybe (?):
"see the following log record for the `app-myvalue-write`..."

The placement of "In that case" (right after the call out description) reads out of context to me. Does it mean ----
"Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML file. In that case......."

Please consider adding a context or removing/revising "In that case....."

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 7]
```diff
@@ -79,32 +79,38 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+  - name: application-logs
+    inputRefs:
+    - application
+    - audit
+    outputRefs:
+    - elasticsearch-secure
+    - default
+    parse: json
+    labels:
+      unmanaged: myvalue
```
Comment (xenolinux) [inline modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:101]: IMO, we should add a call out and a description for this key-value pair.

**Review by xenolinux — COMMENTED** [review]
Added a few comments/suggestions. Let me know if you have questions

/remove-label peer-review-needed
/remove-label peer-review-in-progress
/label peer-review-done

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 8]
```diff
@@ -79,32 +79,38 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
 ----
-<1> Uses the value of the key-value pair that is formed by the OpenShift `myLabel` label.
-<2> The `myLabel` element gives its string value, `myValue`, to the structured log record.
+<1> `outputDefaults` applies only to the default Elasticsearch logstore and not to the other defined Elasticsearch outputs.
```
[the remark this answers is not in the document]
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:103]: Rephrased sentence, "Settings that you specify in the `outputDefaults` field apply only to the default Elasticsearch logstore." looks enough and good to me.

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 9]
```diff
@@ -79,32 +79,38 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+<2> Use the value found at this path in the record to format the index name for the default output.
```
[the remark this answers is not in the document]
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:104]: Uses the value formed by the OpenShift `managed` label in the log record to format the index name for the default output. WDYT?

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc [hunk 10]
```diff
@@ -79,32 +79,38 @@ Suppose that you use the following snippet in your `ClusterLogForwarder` CR YAML
...
+<3> Use the value defined in pipeline to format the index name for the elasticsearch output named `elasticsearch-secure`.
```
[the remark this answers is not in the document]
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: This suggestion is addressed [here](https://github.com/openshift/openshift-docs/pull/75444#discussion_r1646260915).

[the remark this answers is not in the document]
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:105]: Ack.

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 6 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:107]: Maybe we could rephrase as following:

"If you use the following snippet in your `ClusterLogForwarder` CR YAML file, see the following log record for the `app-myvalue-write` index:"

WDYT?

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 7 above)
Comment (smunje1) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:101]: @jcantrill @anpingli Do you think we should add a callout here?

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 6 above)
Comment (xenolinux) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:107]: Sounds good

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (xenolinux) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:104]: Sounds good

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 7 above)
Comment (anpingli) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:101]: yes,  you can. It can be any valid string.

File: modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc (hunk 6 above)
Comment (jcantrill) [reply modules/cluster-logging-configuration-of-json-log-data-for-default-elasticsearch.adoc:107]: the log record doesn't have 'app-myvalue-write' field or even an index so "seeing the following log record is not correct. It should be something like:

The above CR will spec the following log record to be stored in the 'app-myvalue-write' index

**Comment by r2d2rnd** [comment]
Hello @smunje1 ,

I feel this is not explaining well how to configure JSON.

It should be clear what's pretty clear the difference between the configuration needed for LogForwarding as JSON to the internal LogStorage (`default`)
```
outputDefaults: 
  elasticsearch:
    structuredTypeKey: kubernetes.labels.managed 
    structuredTypeName: nologformat
```

And other outputs:
```
outputs:
  - name: elasticsearch-secure
    elasticsearch:
      structuredTypeKey: openshift.labels.unmanaged 
      structuredTypeName: nologformat
```

As it's now, it's not so much clear and it should be.

At the same time, this command is wrong:
```
$ oc delete pod --selector logging-infra=collector
```

For deleting the collector pods, it should be as below, but really, any change in the ClusterLogForward should automatically restart the collector pods, then, the command shouldn't be needed and in case that not restarted, then, it should be reviewed in the ClusterLogForwarder status if any error is present in the logic that leads to don't be restarted automatically:
```
$ oc -n openshift-logging delete pod -l component=collector
```

And the last part, I don't understand the reason for having separated the sections below as "Configuring" and "Forwarding" have the same meaning here. If you configure it, you are log forwarding it, then, what's the sense of having two different sections where even the second "Forwarding..." has not an entire ClusterLogforwarding example and it's only present in the first section "Configuring JSON..."?

 - [Configuring JSON log data for Elasticsearch](https://75444--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/logging/log_collection_forwarding/cluster-logging-enabling-json-logging.html#cluster-logging-configuration-of-json-log-data-for-default-elasticsearch_cluster-logging-enabling-json-logging)
 - [Forwarding JSON logs to the Elasticsearch log store](https://75444--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/logging/log_collection_forwarding/cluster-logging-enabling-json-logging.html#cluster-logging-forwarding-json-logs-to-the-default-elasticsearch_cluster-logging-enabling-json-logging)

**Comment by briandooley** [comment]
@r2d2rnd We would like to deal with [your feedback](https://github.com/openshift/openshift-docs/pull/75444#issuecomment-2189368323) in a separate jira and merge this one, since it has already passed SME and QE review. Does that suit you?

**Comment by r2d2rnd** [comment]
Hello @briandooley ,

My comment has 3 parts. The fist part is directly related to documentation bug OBSDOCS-118 and this is the reason of this ticket. 

For the other two, I'll open a separated issue for it as I understand is out of the scope of this request and also the last part it's more related how this section is structured where it's not so much understandable from a point of view of an user
